### ISSUE TITLE
Serializing a 3D-Array (Type double)

### ISSUE THREAD

**Issue by hobohak** [wojtpl2/ExtendedXmlSerializer#170]
Hello,

there seems to be an issue serializing 3D-array type double.

I set up a simplified example.
Console Application / Install Extension with Install-Package ExtendedXmlSerializer -Version 2.1.1 

**Code Structure**
1. main
2. public class (target for serialization)
3. static helper class to handle se-/deserialization by file

**Logic:** 
Create simple object > Initialize 1D,2D and 3D Array > Serialize 2 file > read from file.

**Error Thrown at Read**
_ExtendedXmlSerializer.Core.Sprache.ParseException: 'An attempt was made to parse the identity 'https://extendedxmlserializer.github.io/system:Double', but no type could be located with that name.'_

**Code**
main
```
class Program
    {
        static void Main(string[] args)
        {
            //filepath for load/save
            string filepath = @"C:\Temp\Person.xml";

            //create tmpobj
            Person tmpP = new Person{Name = "Test Person"};
            tmpP.Initialize();

            //write 2 file
            Helper_Serial.WriteToXmlFileNG<Person>(filepath, tmpP);

            //read 2 file
            Person loadPerson = Helper_Serial.ReadFromXmlFileNG<Person>(filepath);

        }
    }
```

person class
```
[Serializable]
    public class Person
    {
        //Props
        public string Name { get; set; }

        public double[] complex1D { get; set; }
        public double[][] complex2D { get; set; }
        public double[][][] complex3D { get; set; }

        //Method 2 initialize complexArr.
        public void Initialize()
        {
            int _one = 1, two = 2, three = 3;
            complex1D = new double[_one];
            complex2D = new double[_one][];
            complex3D = new double[_one][][];

            for (int i = 0; i < _one; i++)
            {
                complex1D[i] = i;
                complex2D[i] = new double[two];
                complex3D[i] = new double[two][];

                for (int j = 0; j < two; j++)
                {
                    complex2D[i][j] = j;
                    complex3D[i][j] = new double[three];
                    for (int k = 0; k < three; k++)
                        complex3D[i][j][k] = k;
                }

            }
        }
    }
```

**helper serial class**
```
    /// <summary>
    /// Helper Class for Serial. 
    /// </summary>
    public static class Helper_Serial
    {
        //Write
        public static bool WriteToXmlFileNG<T>(string filepath, T ObjectToWrite, bool append = false) where T : new()
        {
            var serializer = new ConfigurationContainer().UseOptimizedNamespaces().Create();
            var xml = serializer.Serialize(ObjectToWrite);
            File.WriteAllText(filepath, serializer.Serialize(ObjectToWrite));
            return true;
        }

        //Read
        public static T ReadFromXmlFileNG<T>(string filepath) where T : new()
        {
            StreamReader reader = null;
            try
            {
                var serializer = new ConfigurationContainer().Create();
                reader = new StreamReader(filepath);
                return serializer.Deserialize<T>(reader);
            }
            catch (Exception) { throw; }
            finally { if (reader != null) { reader.Close(); } }
        }
    }
```

**Person.xml after writeoperation**
```
<?xml version="1.0" encoding="utf-8"?>
<Person xmlns="clr-namespace:XML_3D_Test;assembly=XML_3D_Test">
  <Name>Test Person</Name>
  <complex1D>
    <double xmlns="https://extendedxmlserializer.github.io/system">0</double>
  </complex1D>
  <complex2D>
    <Array xmlns:exs="https://extendedxmlserializer.github.io/v2" exs:item="double" xmlns="https://extendedxmlserializer.github.io/system">
      <double>0</double>
      <double>1</double>
    </Array>
  </complex2D>
  <complex3D>
    <Array xmlns:exs="https://extendedxmlserializer.github.io/v2" exs:item="Double[]" xmlns="https://extendedxmlserializer.github.io/system">
      <Array exs:item="double">
        <double>0</double>
        <double>1</double>
        <double>2</double>
      </Array>
      <Array exs:item="double">
        <double>0</double>
        <double>1</double>
        <double>2</double>
      </Array>
    </Array>
  </complex3D>
</Person>
```

Any help or advise appreciated. Thank you for that great product though.

**Comment by Mike-EEE** [comment]
Wow that is amazing that it made it that far. 😆 ... That is definitely one type that I don't think we've ever expected to support.  I will definitely look into this.  Do you know if it's supported by the classic XmlSerializer?  I wonder if Xaml would serialize this, too...

Looks like we're serializing OK, just not deserializing.

**Comment by Mike-EEE** [comment]
Doesn't look like the classic serializer supports this... but is that a surprise? ;)

https://stackoverflow.com/questions/30798205/net-xmlserializer-and-multidimensional-arrays

**Comment by hobohak** [comment]
Quick answer \o/

Yes it does -> No error thrown with those static functions
```
        public static bool WriteToXmlFile<T>(string filePath, T ObjectToWrite, bool append = false) where T : new()
        {
            TextWriter writer = null;
            var serializer = new XmlSerializer(typeof(T));
            writer = new StreamWriter(filePath, append);
            serializer.Serialize(writer, ObjectToWrite);
            if (writer != null)
                writer.Close();

            return true;

        }

        public static T ReadFromXmlFile<T>(string filePath) where T : new()
        {
            TextReader reader = null;
            try
            {
                var serializer = new XmlSerializer(typeof(T));
                reader = new StreamReader(filePath);
                return (T)serializer.Deserialize(reader);
            }
            finally
            {
                if (reader != null)
                    reader.Close();
            }
        }
```

**Comment by Mike-EEE** [comment]
Haha... oh wow.  That IS surprising.  Guess I misunderstood the StackOverflow question as it seemed it wasn't possible/supported.  I did just clear some time to check it out and yes indeed it works as advertised.  No hiding now. 😉  I will be looking into this.

**Comment by hobohak** [comment]
Thank you :+1:

**Comment by Mike-EEE** [comment]
Aha!  I figured out my disconnect.  What we are serializing above is actually a [jagged array](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/jagged-arrays), not a [multi-dimensional array](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/arrays/multidimensional-arrays).  Multi-dimensional arrays are not supported by XmlSerializer, while jagged arrays are.  Which, the irony of course is that multi-dimensional are of fixed dimensions and format so should be easier to emit. 😉

**Comment by Mike-EEE** [comment]
OK @hobohak ... got first draft support for jagged arrays for you.  You can try them out here:

https://www.myget.org/F/wojtpl2/api/v3/index.json

Multidimensional arrays do not work yet but I should have them by EOD.  We'll see how I do. ;)

**Comment by hobohak** [comment]
Hey @Mike-EEE thanks for your efforts.

I have included the myget feed and installed ExtSer. from there (v2.1.2 (latest stable). More up to date options were not available. What am I doing wrong?

Have modified the code slightly for MultiArray to test but not possible b/c of that (at least I think :) )


**Adapted main**
```
        static void Main(string[] args)
        {
            //filepath for load/save
            string filepathMulti = @"C:\Temp\PersonMulti.xml";
            string filepathJagged = @"C:\Temp\PersonPersonJagged.xml";

            //create tmpobj
            Person_Multi tmpMulti = new Person_Multi { Name = "Test Person" };
            Person_Jagged tmpJagged = new Person_Jagged { Name = "Test Person" };
            tmpMulti.Initialize();
            tmpJagged.Initialize();

            //write 2 file
            Helper_Serial.WriteToXmlFile<Person_Multi>(filepathMulti, tmpMulti);
            Helper_Serial.WriteToXmlFile<Person_Jagged>(filepathJagged, tmpJagged);

            //read 2 file
            Person_Multi loadPersonMulti = Helper_Serial.ReadFromXmlFile<Person_Multi>(filepathMulti);
            Person_Jagged loadPersonJagged = Helper_Serial.ReadFromXmlFile<Person_Jagged>(filepathJagged);

        }
```

**Person Multi**
```
    [Serializable]
    public class Person_Multi
    {
        int one = 1, two = 2, three = 3;
        //Props
        public string Name { get; set; }

        public double[] complex1D { get; set; }
        public double[,] complex2D { get; set; }
        public double[,,] complex3D { get; set; }

        //Method 2 initialize complexArr.
        public void Initialize()
        {

            complex1D = new double[one];
            complex2D = new double[one,two];
            complex3D = new double[one,two,three];

            for (int i = 0; i < one; i++)
            {
                complex1D[i] = i;

                for (int j = 0; j < two; j++)
                {
                    complex2D[i,j] = j;
                    for (int k = 0; k < three; k++)
                        complex3D[i,j,k] = k;
                }

            }
        }
    }
```

**Comment by Mike-EEE** [comment]
OK I just added support for multi-2D.  Only because there was [an example of how to do it](https://stackoverflow.com/a/1183019/3602057). 😉 There doesn't seem to be a lot of support out there around this scenario, as "multidimensional" ends up with 2D, and not 3D.  Of course, there are up to 32-D if necessary.  So I just kept it to 2D for now, and made it easy to add 3D (or more).

If you are interested in adding 3D (or more) support, it would require 3 things (most of it copy/paste):

1. [Adding it to the list of known dimension mappers](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/93220ac8293f75dca0c2df1765fcff9950b02510/src/ExtendedXmlSerializer/ContentModel/Collections/DimensionMappers.cs#L38). These are indexed by the number of dimensions.  Currently there is only one that is registered with `2` that handles the 2D.  So in the case of 3D you would copy the entry (changing 2D to 3D of course) and update the entry to `3`.
1. [Creating a Generic bridge for the mapper](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/93220ac8293f75dca0c2df1765fcff9950b02510/src/ExtendedXmlSerializer/ContentModel/Collections/DimensionMapper2D.cs).  This is how we get into generic land and can do generic things. 😄 You can see that the source array and the dimensions of the instance are passed in.
1. [Creating the actual mapping implementation](https://github.com/wojtpl2/ExtendedXmlSerializer/blob/93220ac8293f75dca0c2df1765fcff9950b02510/src/ExtendedXmlSerializer/ContentModel/Collections/Mapping2D.cs).  Pretty straight forward.

If you (or someone) can submit a PR (along with passing a test) that does this for 3D, I will add it in for you. 😄

**Comment by Mike-EEE** [comment]
Also, as for the feed, there is only one version that is available, the latest.  So what you installed was the latest.  If you do a refresh, you should see an update with the latest changes.

**Comment by hobohak** [comment]
Ok, thank you. I'll check in spare time for the pr.

**Comment by Mike-EEE** [comment]
Cool... lemme know if you run into any problems and I'll help you out. 👍

**Comment by Mike-EEE** [comment]
Hey @hobohak not to be fickle. :)  But I was thinking of this and it should be really easy to implement.  Hopefully you haven't started on this and I can have something for you soon.

**Comment by Mike-EEE** [comment]
Hey @hobohak check out the latest version now... It supports **ALL** 32-dimension arrays. 😄

**Comment by hobohak** [comment]
Haha! :) No I wanted to shuffle on the weekend! But you were quicker :+1: 

Think I found something else though. If I have a  an element like a view in an interface. If inherited classes have a more specific viewelement like a label with a concrete parser it seems to serialize correctly, but not reverse.

I will check, if I can reproduce in simple example and post as a different topic, all right?

**Comment by hobohak** [comment]
Ok, I checkd the build from ur myget feed. 
And it fails the following test Case:

using Microsoft.VisualStudio.TestTools.UnitTesting;
using UnitTestXML.Helper;
using static UnitTestXML.Helper.DeSerialize;
**Result**
- Jagged Passes
- Multi fails
---------------
ErrorMessage:
_Message: Test method UnitTestXML.ExtendenXMLTestLocal.MultiArray threw exception: 
System.InvalidCastException: Das Objekt des Typs "System.Double[]" kann nicht in Typ "System.Double[,]" umgewandelt werden._

```cs
namespace UnitTestXML
{
    [TestClass]
    public class ExtendenXMLTestLocal
    {
        [TestMethod]
        public void MultiArray()
        {
            //filepath for load/save
            string filepathMulti = @"C:\Temp\PersonMulti.xml";

            //create tmpobj
            Person_Multi tmpMulti = new Person_Multi { Name = "Multi Person" };
            tmpMulti.Initialize();

            //write 2 file
            Helper_Serial.WriteToXmlFileNG<Person_Multi>(filepathMulti, tmpMulti);

            //read 2 file
            Person_Multi loadPersonMulti = Helper_Serial.ReadFromXmlFileNG<Person_Multi>(filepathMulti);
        }

        [TestMethod]
        public void JaggedArray()
        {
            //filepath for load/save
            string filepathJagged = @"C:\Temp\PersonPersonJagged.xml";

            //create tmpobj
            Person_Jagged tmpJagged = new Person_Jagged { Name = "Jagged Person" };
            tmpJagged.Initialize();

            //write 2 file
            Helper_Serial.WriteToXmlFileNG<Person_Jagged>(filepathJagged, tmpJagged);

            //read 2 file
            Person_Jagged loadPersonJagged = Helper_Serial.ReadFromXmlFileNG<Person_Jagged>(filepathJagged);
        }
    }
}
```

**TestClasses used:**
```cs
using System;

namespace UnitTestXML.Helper
{
    [Serializable]
    public class Person_Jagged
    {
        //Props
        public string Name { get; set; }

        public double[] complex1D { get; set; }
        public double[][] complex2D { get; set; }
        public double[][][] complex3D { get; set; }

        //Method 2 initialize complexArr.
        public void Initialize()
        {
            int _one = 1, two = 2, three = 3;
            complex1D = new double[_one];
            complex2D = new double[_one][];
            complex3D = new double[_one][][];

            for (int i = 0; i < _one; i++)
            {
                complex1D[i] = i;
                complex2D[i] = new double[two];
                complex3D[i] = new double[two][];

                for (int j = 0; j < two; j++)
                {
                    complex2D[i][j] = j;
                    complex3D[i][j] = new double[three];
                    for (int k = 0; k < three; k++)
                        complex3D[i][j][k] = k;
                }
            }
        }
    }

    [Serializable]
    public class Person_Multi
    {
        int one = 1, two = 2, three = 3;
        //Props
        public string Name { get; set; }

        public double[] complex1D { get; set; }
        public double[,] complex2D { get; set; }
        public double[,,] complex3D { get; set; }

        //Method 2 initialize complexArr.
        public void Initialize()
        {
            complex1D = new double[one];
            complex2D = new double[one, two];
            complex3D = new double[one, two, three];

            for (int i = 0; i < one; i++)
            {
                complex1D[i] = i;

                for (int j = 0; j < two; j++)
                {
                    complex2D[i, j] = j;
                    for (int k = 0; k < three; k++)
                        complex3D[i, j, k] = k;
                }

            }
        }
    }
}

```

**Helpers (for DeSerial..)**
```cs
using ExtendedXmlSerializer.Configuration;
using ExtendedXmlSerializer.ExtensionModel.Xml;
using System;
using System.IO;
using System.Xml.Serialization;

namespace UnitTestXML.Helper
{
    public class DeSerialize
    {
        public static class Helper_Serial
        {
            //Write Extended
            public static bool WriteToXmlFileNG<T>(string filepath, T ObjectToWrite, bool append = false) where T : new()
            {
                var serializer = new ConfigurationContainer().UseOptimizedNamespaces().Create();
                var xml = serializer.Serialize(ObjectToWrite);
                File.WriteAllText(filepath, serializer.Serialize(ObjectToWrite));
                return true;
            }

            //Read Extended
            public static T ReadFromXmlFileNG<T>(string filepath) where T : new()
            {
                StreamReader reader = null;
                try
                {
                    var serializer = new ConfigurationContainer().UseOptimizedNamespaces().Create();
                    reader = new StreamReader(filepath);
                    return serializer.Deserialize<T>(reader);
                }
                catch (Exception) { throw; }
                finally { if (reader != null) { reader.Close(); } }
            }

            //Write Original XML
            public static bool WriteToXmlFile<T>(string filePath, T ObjectToWrite, bool append = false) where T : new()
            {
                TextWriter writer = null;
                var serializer = new XmlSerializer(typeof(T));
                writer = new StreamWriter(filePath, append);
                serializer.Serialize(writer, ObjectToWrite);
                if (writer != null)
                    writer.Close();

                return true;

            }

            //Read Original 
            public static T ReadFromXmlFile<T>(string filePath) where T : new()
            {
                TextReader reader = null;
                try
                {
                    var serializer = new XmlSerializer(typeof(T));
                    reader = new StreamReader(filePath);
                    return (T)serializer.Deserialize(reader);
                }
                finally
                {
                    if (reader != null)
                        reader.Close();
                }
            }

        }
    }
}
```

The XML from MultiPerson:
```xml
<?xml version="1.0" encoding="utf-8"?>
<Person_Multi xmlns="clr-namespace:UnitTestXML.Helper;assembly=UnitTestXML">
  <Name>Test Person</Name>
  <complex1D>
    <double xmlns="https://extendedxmlserializer.github.io/system">0</double>
  </complex1D>
  <complex2D>
    <double xmlns="https://extendedxmlserializer.github.io/system">0</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">1</double>
  </complex2D>
  <complex3D>
    <double xmlns="https://extendedxmlserializer.github.io/system">0</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">1</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">2</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">0</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">1</double>
    <double xmlns="https://extendedxmlserializer.github.io/system">2</double>
  </complex3D>
</Person_Multi>
```

**Comment by Mike-EEE** [comment]
Boh!  It's not emitting the mapping information in members.  I'll look into that now. 

> I will check, if I can reproduce in simple example and post as a different topic, all right?

Yeah for sure if you find another issue post the data like you have here (in another issue would probably be best!) and we'll see if we can figure out. 😄

**Comment by Mike-EEE** [comment]
Alright @hobohak ... got a fix in for you.  Try it now. 🤞

**Comment by hobohak** [comment]
Sorry, have been sick fo a couple of days. Have tested the code with n-multidimensional array and it passes the tests. I am using code contributed bt Saeeduddin on so / see below to Loop through n-Array and fill up with values (slightly adapted). With this I was able to randomly to more testcases and they all pass.


**Code from Saeeduddin to fill multiArray**
```cs
    /// <summary>
    /// Credit to: Asad Saeeduddin
    /// https://stackoverflow.com/questions/28867327/looping-through-a-n-dimensional-array#28868622
    /// </summary>
    public static class Helper_MultiDimension
    {
        public static int[] IndexToCoordinates(int i, Array arr)
        {
            var dims = Enumerable.Range(0, arr.Rank)
                .Select(arr.GetLength)
                .ToArray();

            Func<int, int, int> product = (i1, i2) => i1 * i2;

            return dims.Select((d, n) => (i / dims.Take(n).Aggregate(1, product)) % d).ToArray();
        }

        public static void OutputAllArrayIndices(Array arr)
        {
            var i = 0;
            int k = 0;

            foreach (double item in arr)
            {
                var coords = IndexToCoordinates(i++, arr);
                arr.SetValue((double)k, coords);

                //System.Diagnostics.Debug.WriteLine(string.Join(", ", coords) + "\t=>" + arr.GetValue(coords));
                k++;
            }
        }
    }
```

**Adapted Initialization**
```cs
    [Serializable]
    public class Person_Multi
    {
        //Props
        public string Name { get; set; }

        public object ComplexND { get; set; }

        public void Initialize()
        {
            var tmpComplex = new double[1, 2, 3, 1, 2, 3];

            Helper_MultiDimension.OutputAllArrayIndices(tmpComplex);
            ComplexND = tmpComplex;

            //var Complex = new double[1, 2, 3, 4, 5, 6,7, 8, 9, 10,11,12,13,14,15,16];
        }

    }
```


**output file**
```xml
<?xml version="1.0" encoding="utf-8"?>
<Person_Multi xmlns:exs="https://extendedxmlserializer.github.io/v2" xmlns:sys="https://extendedxmlserializer.github.io/system" xmlns="clr-namespace:UnitTestXML.Helper;assembly=Test">
  <Name>Multi Person</Name>
  <ComplexND exs:type="sys:double^6" exs:map="1,2,3,1,2,3">
    <sys:double>0</sys:double>
    <sys:double>12</sys:double>
    <sys:double>24</sys:double>
    <sys:double>6</sys:double>
    <sys:double>18</sys:double>
    <sys:double>30</sys:double>
    <sys:double>2</sys:double>
    <sys:double>14</sys:double>
    <sys:double>26</sys:double>
    <sys:double>8</sys:double>
    <sys:double>20</sys:double>
    <sys:double>32</sys:double>
    <sys:double>4</sys:double>
    <sys:double>16</sys:double>
    <sys:double>28</sys:double>
    <sys:double>10</sys:double>
    <sys:double>22</sys:double>
    <sys:double>34</sys:double>
    <sys:double>1</sys:double>
    <sys:double>13</sys:double>
    <sys:double>25</sys:double>
    <sys:double>7</sys:double>
    <sys:double>19</sys:double>
    <sys:double>31</sys:double>
    <sys:double>3</sys:double>
    <sys:double>15</sys:double>
    <sys:double>27</sys:double>
    <sys:double>9</sys:double>
    <sys:double>21</sys:double>
    <sys:double>33</sys:double>
    <sys:double>5</sys:double>
    <sys:double>17</sys:double>
    <sys:double>29</sys:double>
    <sys:double>11</sys:double>
    <sys:double>23</sys:double>
    <sys:double>35</sys:double>
  </ComplexND>
</Person_Multi>
```

**Comment by Mike-EEE** [comment]
Woohoo!  That's awesome.  TBH I am amazed it even works, haha.  Figured there must be a reason why it wasn't supported in the classic XmlSerializer.  I got to learn a little bit about arrays to boot. 😉 

For version 3 I am thinking of enabling formatting support for arrays that have elements with a registered converter, so that it will emit them as one string instead of many elements.   for example, the array above would be something like:

```
<ComplexND exs:type="sys:double^6" exs:map="1,2,3,1,2,3">0,12,24,6,18,30,2,14,26,8,20,32,4,16,28,10,22,34,1,13,25,7,19,31,3,15,27,9,21,33,5,17,29,11,23,35</ComplexND>
```

We actually already do something similar to this for our `byte[]` arrays, but not others.  Should be easy enough to extend.  I would introduce it to v2 but it's pretty much in quality release mode while we work on v3 (total rewrite).
